### PR TITLE
Update libp2p to v0.3

### PR DESCRIPTION
**Pull request by tomaka**
- Updates `ring` to 0.14.
- Moves the topology into the `custom_protos` module (see https://github.com/libp2p/rust-libp2p/pull/889)

I'm marking this as "WIP" because the connectivity test has some issues which I'm trying to find out.
We spawn 25 nodes and the test succeeds when all the nodes are connected to the 24 others, but there are sometimes 2 or 3 them stuck at 22 or 23 connections.

### CODE REVIEW

**Comment by tomaka** [comment]
For what it's worth, the test failure is very very likely caused by nodes trying to dial each other simultaneously. This is something that in practice only happens in a local network.
In `--release`, the test passes.
However I'd like to get to the bottom of this and try to properly fix this problem once and for all.

**Comment by arkpar** [comment]
Does it break compatibility?

**Comment by tomaka** [comment]
No. It is 0.3 because of some API breaking changes within libp2p itself.

**Comment by arkpar** [comment]
I mean the network level compatibility. Would it be able to connect and communicate to 0.2 nodes?

**Comment by tomaka** [comment]
Yes!

File: core/network-libp2p/src/service_task.rs [hunk 1]
```diff
@@ -90,11 +73,15 @@ where TProtos: IntoIterator<Item = RegisteredProtocol> {
 		}
 	}
 
-	// Add the bootstrap nodes to the topology and connect to them.
+	/*// Add external addresses.
```
Comment (bkchr) [inline core/network-libp2p/src/service_task.rs]: Is that still required or just some leftover?

Comment (tomaka) [reply core/network-libp2p/src/service_task.rs]: Oh, I forgot to uncomment this.

File: core/network-libp2p/src/custom_proto/handler.rs [hunk 2]
```diff
@@ -140,11 +146,17 @@ where
...
 		match self.state {
-			State::Disabled | State::ShuttingDown => return,
-			State::Normal => ()
+			// TODO: Normally we should refuse incoming connections if we're disabled; however
+			// tests are failing if we do that because on localhost substreams arrive quicker than
+			// we receive the "Enabled" message from the network behaviour. Which fix to employ is
+			// kind of blurry. This problem will become irrelevant after
+			// https://github.com/paritytech/substrate/issues/1517
+			// TODO: also, we should shut down refused substreams gracefully; this should be fixed
+			// at the same time as the todo above
```
Comment (arkpar) [inline core/network-libp2p/src/custom_proto/handler.rs]: Why not just start in `Enabled` state?

File: core/network-libp2p/src/service_task.rs [hunk 3]
```diff
@@ -90,11 +73,15 @@ where TProtos: IntoIterator<Item = RegisteredProtocol> {
...
+	for addr in &config.public_addresses {
+		Swarm::add_external_address(addr.clone());
+	}*/
```
Comment (arkpar) [inline core/network-libp2p/src/service_task.rs]: Commented out code should be explained or removed

**Pull request merged**